src/uct/ib/base/ib_iface.c

```diff
@@ -420,30 +422,41 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
         {
             ucs_error("ibv_query_pkey("UCT_IB_IFACE_FMT", index=%d) failed: %m",
                       UCT_IB_IFACE_ARG(iface), pkey_index);
+            continue;
         }
 
         pkey = ntohs(port_pkey);
         if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
             /* if pkey = 0x0, just skip it w/o debug trace, because 0x0
              * means that there is no real pkey configured at this index */
             if (pkey) {
-                ucs_debug("skipping send-only pkey[%d]=0x%x", pkey_index, pkey);
+                ucs_trace("skipping send-only pkey[%d]=0x%x on "UCT_IB_IFACE_FMT,
+                          pkey_index, pkey, UCT_IB_IFACE_ARG(iface));
             }
             continue;
         }
 
         /* take only the lower 15 bits for the comparison */
-        if (!pkey_already_set ||
+        if ((config->pkey_value == UCS_HEXUNITS_AUTO) ||
             ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value)) {
             iface->pkey_index = pkey_index;
             iface->pkey_value = pkey;
-            pkey_already_set  = 1;
+            pkey_set          = 1;
+            break;
         }
     }
 
-    if (!pkey_already_set) {
-        ucs_error("There is no valid pkey with full membership on %s:%d",
-                  uct_ib_device_name(dev), iface->config.port_num);
+    if (!pkey_set) {
+        if (config->pkey_value == UCS_HEXUNITS_AUTO) {
+            ucs_error("There is no valid pkey with full membership on "
+                      UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+        } else {
+            ucs_config_sprintf_hex(pkey_str, UCT_IB_PKEY_STR_MAX,
```

Comment:
no need to define UCT_IB_PKEY_STR_MAX and pkey_str - just use "0x%x" in the ucs_error below

Reply:
good catch
done